File: hw06/models/animals/pets/Cat.java
package hw06.models.animals.pets;

import java.util.Date;
import java.util.List;

/**
 * Cat is a class inherited from classes Animal -> Pet
 */
public class Cat extends Pet{
    private boolean woolly;

    /**
     * It is a constructor for Cat class
     * @param height            The height of animal
     * @param weight            The weight of animal
     * @param eyeColor          The eye color of animal
     * @param name              The name of animal
     * @param breed             The breed of animal
     * @param vaccinations      The vaccinations list of animal
     * @param furColor          The fur color of animal
     * @param birthDate         The birthdate of animal
     * @param woolly            The woolly status of animal
     */
    public Cat(double height, double weight, String eyeColor, String name, String breed, List<String> vaccinations, String furColor, Date birthDate, boolean woolly) {
        super(height, weight, eyeColor, name, breed, vaccinations, furColor, birthDate);
        this.woolly = woolly;
    }
    
    /**
     * It is a constructor for Cat class (woolly status will be false)
     * @param height            The height of animal
     * @param weight            The weight of animal
     * @param eyeColor          The eye color of animal
     * @param name              The name of animal
     * @param breed             The breed of animal
     * @param vaccinations      The vaccinations list of animal
     * @param furColor          The fur color of animal
     * @param birthDate         The birthdate of animal
     * @param woolly            The woolly status of animal
     */
    public Cat(double height, double weight, String eyeColor, String name, String breed, List<String> vaccinations, String furColor, Date birthDate) {
        super(height, weight, eyeColor, name, breed, vaccinations, furColor, birthDate);
        this.woolly = false;
    }

    /**
     * Returns is the animal woolly or not
     * @return this.woolly
     */
    public boolean isWoolly() {
        return this.woolly;
    }

    @Override
    public String sound() {
        return "Meow-meow!";
    }

    @Override
    public String toString() {
        return String.format("Cat (%s, is woolly: %b)", super.toString(), this.isWoolly());
    }
}

File: hw07/internal/models/cat.go
package models

import (
	"fmt"
	"time"
)

// Cat structure
type Cat struct {
	Pet
	woolly bool
}

// Creates new Cat instance with full parameters given
func NewCat(height float64, weight float64, eyeColor string, name string, breed string, vaccinations []string, furColor string, birthDate time.Time, woolly bool) (*Cat, error) {
	p, err := newPet(height, weight, eyeColor, name, breed, vaccinations, furColor, birthDate)
	if err != nil {
		return nil, err
	}

	c := &Cat{
		*p,
		woolly,
	}

	// Handles potential restrictions of any parameter to return error when new Cat cannot be instantiated
	return c, nil
}

// Returns the is cat woolly
func (c *Cat) IsWoolly() bool {
	return c.woolly
}

// Returns sound of cat
func (c *Cat) Sound() string {
	return fmt.Sprint("Meow-Meow")
}

// Returns the string representation of Cat instance
func (c *Cat) String() string {
	return fmt.Sprintf("Cat %v, is woolly: %t", c.Pet.String(), c.woolly)
}
